### PR TITLE
Update neo4j-driver

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-17T19:22:26Z" title="Tuesday, December 17th 2019, 8:22:26 pm +01:00">Dec 17, 2019</time>_
_Merged <time datetime="2020-01-07T11:31:13Z" title="Tuesday, January 7th 2020, 12:31:13 pm +01:00">Jan 7, 2020</time>_
---

Fix API changes. Also close the session in resolver only and refactor
userMiddleware to become a part of the resolver.

The fix for the test was to move `session.close` into the finally block at the end of the resolver. It was complaining about using a session in `userMiddleware.js` which was already closed.

### CODE REVIEW

**Comment by Tirokk** [comment]
> [cypress[bot]](https://github.com/apps/cypress) **Authored by [cypress[bot]](https://github.com/apps/cypress)**
_<time datetime="2019-12-17T20:30:10Z" title="Tuesday, December 17th 2019, 9:30:10 pm +01:00">Dec 17, 2019</time>_

---



<p align="center">
  <br />
  
</p>

---

### Test summary
<a href="https://dashboard.cypress.io/projects/qa7fe2/runs/3604/specs"></a> <b title="Passed">50</b>  &bull; <a href="https://dashboard.cypress.io/projects/qa7fe2/runs/3604/failures"></a> <b title="Failed">0</b>  &bull;  <b title="Pending">0</b> &bull;  <b title="Skipped">0</b>

---

### Run details
| | |
| --- | --- |
| **Project** | Human-Connection |
| **Status** | Passed |
| **Commit** | a568c96c5e |
| **Started** | Jan 7, 2020 11:07 AM |
| **Ended** | Jan 7, 2020 11:18 AM |
| **Duration** | 11:37 <a href="https://on.cypress.io/parallelization" title="Enable parallelization to reduce run duration">💡</a> |
| **OS** | Linux Ubuntu Linux - 16.04 |
| **Browser** | Chromium 79  |

[View run in Cypress Dashboard ➡️](https://dashboard.cypress.io/projects/qa7fe2/runs/3604/specs)



---

This comment has been generated by **cypress-bot** as a result of this project's GitHub integration settings. You can manage this integration in [this project's settings in the Cypress Dashboard](https://dashboard.cypress.io/projects/qa7fe2/settings)

File: backend/src/middleware/user/userMiddleware.js [hunk 1]
```diff
@@ -1,16 +0,0 @@
...
-      await createOrUpdateLocations(result.id, args.locationName, context.driver)
-      return result
-    },
-    UpdateUser: async (resolve, root, args, context, info) => {
-      const result = await resolve(root, args, context, info)
-      await createOrUpdateLocations(args.id, args.locationName, context.driver)
-      return result
-    },
-  },
-}
```
Comment (Tirokk) [inline backend/src/middleware/user/userMiddleware.js:16]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-19T14:32:26Z" title="Thursday, December 19th 2019, 3:32:26 pm +01:00">Dec 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,16 +0,0 @@
-import createOrUpdateLocations from '../nodes/locations'
-
-export default {
-  Mutation: {
-    SignupVerification: async (resolve, root, args, context, info) => {
-      const result = await resolve(root, args, context, info)
-      await createOrUpdateLocations(result.id, args.locationName, context.driver)
-      return result
-    },
-    UpdateUser: async (resolve, root, args, context, info) => {
-      const result = await resolve(root, args, context, info)
-      await createOrUpdateLocations(args.id, args.locationName, context.driver)
-      return result
-    },
-  },
-}
```


why are these changes in this `PR` @roschaefer 
code creep?

File: backend/src/schema/resolvers/registration.js [hunk 2]
```diff
@@ -51,11 +54,14 @@ export default {
...
           emailAddress.update({ verifiedAt: new Date().toISOString() }),
         ])
+        await createOrUpdateLocations(args.id, args.locationName, session)
         return user.toJson()
       } catch (e) {
         if (e.code === 'Neo.ClientError.Schema.ConstraintValidationFailed')
           throw new UserInputError('User with this slug already exists!')
         throw new UserInputError(e.message)
+      } finally {
+        session.close()
```
Comment (Tirokk) [inline backend/src/schema/resolvers/registration.js:64]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-19T14:33:14Z" title="Thursday, December 19th 2019, 3:33:14 pm +01:00">Dec 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -51,11 +54,14 @@ export default {
           emailAddress.relateTo(user, 'belongsTo'),
           emailAddress.update({ verifiedAt: new Date().toISOString() }),
         ])
+        await createOrUpdateLocations(args.id, args.locationName, session)
         return user.toJson()
       } catch (e) {
         if (e.code === 'Neo.ClientError.Schema.ConstraintValidationFailed')
           throw new UserInputError('User with this slug already exists!')
         throw new UserInputError(e.message)
+      } finally {
+        session.close()
```


maybe these make sense here, if so, please explain... 
if not, please open an issue to refactor, then add them there where our future selves can find them more easily

File: backend/src/schema/resolvers/users/location.js [hunk 3]
```diff
@@ -110,44 +105,36 @@ const createOrUpdateLocations = async (userId, locationName, driver) => {
...
+          MATCH (location:Location {id: $locationId})
+          MERGE (user)-[:IS_IN]->(location)
           RETURN location.id, user.id
         `,
-        { userId: userId, locationId: data.id },
-      )
-    })
-  } finally {
-    session.close()
-  }
```
Comment (Tirokk) [inline backend/src/schema/resolvers/users/location.js:150]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-19T14:33:51Z" title="Thursday, December 19th 2019, 3:33:51 pm +01:00">Dec 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -110,44 +105,36 @@ const createOrUpdateLocations = async (userId, locationName, driver) => {
   if (data.context) {
     await asyncForEach(data.context, async ctx => {
       await createLocation(session, ctx)
-      try {
-        await session.writeTransaction(transaction => {
-          return transaction.run(
-            `
+      await session.writeTransaction(transaction => {
+        return transaction.run(
+          `
               MATCH (parent:Location {id: $parentId}), (child:Location {id: $childId})
               MERGE (child)<-[:IS_IN]-(parent)
               RETURN child.id, parent.id
             `,
-            {
-              parentId: parent.id,
-              childId: ctx.id,
-            },
-          )
-        })
-        parent = ctx
-      } finally {
-        session.close()
-      }
+          {
+            parentId: parent.id,
+            childId: ctx.id,
+          },
+        )
+      })
+      parent = ctx
     })
   }
   // delete all current locations from user and add new location
-  try {
-    await session.writeTransaction(transaction => {
-      return transaction.run(
-        `
+  await session.writeTransaction(transaction => {
+    return transaction.run(
+      `
           MATCH (user:User {id: $userId})-[relationship:IS_IN]->(location:Location)
           DETACH DELETE relationship
           WITH user
-          MATCH (location:Location {id: $locationId}) 
-          MERGE (user)-[:IS_IN]->(location) 
+          MATCH (location:Location {id: $locationId})
+          MERGE (user)-[:IS_IN]->(location)
           RETURN location.id, user.id
         `,
-        { userId: userId, locationId: data.id },
-      )
-    })
-  } finally {
-    session.close()
-  }
```


why are you no longer closing the session here in this function?
is it closed later?

File: backend/src/schema/resolvers/registration.js (hunk 2 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/registration.js:64]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-19T14:44:38Z" title="Thursday, December 19th 2019, 3:44:38 pm +01:00">Dec 19, 2019</time>_

---


The signature of `createOrUpdateLocations` has changed. It is now expecting session and not driver. When I updated I got complaints that a session was already closed. So I re-use the same session and close it at the end.

File: backend/src/schema/resolvers/users/location.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/users/location.js:150]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-12-19T14:45:22Z" title="Thursday, December 19th 2019, 3:45:22 pm +01:00">Dec 19, 2019</time>_

---


See above :point_up:

**Review by Mogge — CHANGES_REQUESTED** [review]
It makes it harder to review when there are unrelated changes @roschaefer 
please split them :smile_cat:

**Review by Mogge — APPROVED** [review]

**Pull request closed without merging**